Add next steps for Crescend get-started (#30)
MicrosoftDocs/PowerShell-Docs-Modules: reference/docs-conceptual/Crescendo/get-started/choose-command-line-tool.md

---
description: How to select a command-line tool to amplify using Crescendo.
ms.date: 03/08/2022
title: Choosing a command-line tool to amplify
---
# Choosing the command-line tool for Crescendo

> Developing a cmdlet with Crescendo is a rapid, reliable way to _amplify_ a command-line tool to
> produce a cmdlet-like experience. Many times, the tool you need to work with is directly tied to the

Reviewer: ~~Been thinking about this a little bit during the broader review; I think it is worth explicitly calling out that the end product is a cmdlet _but_ "developing a cmdlet... to produce a cmdlet-like experience" feels a little off. This also introduces the "developing a cmdlet" versus "amplifying a command-line tool" terminology distinction I mention elsewhere. I keep trying to rephrase this line without losing the point of it (from my perspective, my key take away as a reader is "Crescendo helps me rapidly and reliably _amplify_ a command-line tool to a cmdlet-like experience"). My best effort:~~
```suggestion
Using Crescendo is a rapid, reliable way to _amplify_ a command-line tool to
produce a cmdlet-like experience. Many times, the tool you need to work with is directly tied to the
```

Edit: Stricken in favor of [this comment]() but left for clarity (and in case the decision is made to update terminology).

technology you are attempting to automate. This makes the tool choice simple. However, not all tools
work the same way. To optimize your investment when developing a cmdlet, consider the following
criteria:

## Command-line tools that make good candidates

- **The original tool is difficult to use.**

  If the tool is simple to use and provides the information you need, there is no need to create a

>   cmdlet. However, this isn't always the case with all command-line tools. Many command-line tools

Reviewer: Implied by "always the case" I think
```suggestion
  cmdlet. However, this isn't always the case with command-line tools. Many command-line tools
```

>   have their own unique syntax, parameters, and output that make it difficult for inexperienced

Reviewer: Nit: `inexperienced` or `unfamiliar`?

>   administrators to use the command. Converting the command into a cmdlet provides all the benefits

Reviewer: My understanding of verbal conversation is that we want to refer to the _thing to be amplified_ as the _tool_ which has one or more _commands_. I'm not sure if the first instance of `command` on this line should be `tool's commands` or `tool`; including options for both ways. I think the distinction I want to make sure we clarify is between `tool`, `command`, `amplified command`, and `cmdlet`.

### Option One

```suggestion
  administrators to use the tool. Converting the tool into a cmdlet provides all the benefits
```

### Option Two

```suggestion
  administrators to use the tool's commands. Converting a tool's command into a cmdlet provides all the benefits
```

Author: Some definitions:
- tool - the executable
- command - the use of the tool on the command line (i.e. tool+cmds+params)
- amplified command - a command wrapped in a cmdlet

Author: I like option 2 but we don't need to say "tool's commands" in the second sentence. We can just say "command".

Reviewer: Agreed - added a third option based on these points.

  of cmdlet discovery, syntax consistency, and structured output as objects.

- **The command-line tool output is difficult to use in automation.**

  Command-line tools output their information to the screen as string data. This is not the
  structured data (objects) that PowerShell expects in the pipeline, which prevents you from using
  PowerShell cmdlets, such as `Where-Object` and `ForEach-Object`. Crescendo assists you with

>   _amplifying_ the command-line tool experience to a cmdlet so that the tool can participate in the

Reviewer: I'm not sure on the phrasing here, something about the rhythm felt off when I said it out loud. Do we want to highlight that Crescendo is providing a cmdlet that amplifies the command-line tool, that Crescendo's amplification _is_ a cmdlet, that Crescendo is specifically amplifying a command-line tool for pipeline functionality, or something else?

```suggestion
  _amplifying_ the command-line tool experience so that the tool can participate in the
```

  PowerShell pipeline.

- **The command-line tool doesn't provide adequate help.**

  Command-line tools that lack help information can be difficult to use. Cmdlets can have
  comprehensive help information that includes details on parameters, descriptions, and examples.
  Crescendo gives you the ability to create the missing help for your amplified commands.

## Command-line tools that make bad candidates

- **Does a cmdlet already exist?**

  PowerShell has a well-established ecosystem of modules that extend the built-in capabilities of
  PowerShell. Before deciding to amplify a command-line tool with Crescendo, first check to see if a
  cmdlet already exists that performs your intended goal. For example, wrapping the Windows command
  `ipconfig.exe` to get the current IP address is a poor choice because the module **NetAdapter**
  contains `Get-NetIPConfiguration`, which provides the same information with structured output.

- **Are there better ways?**

  Often the information you need for automation can be obtained quicker than making a cmdlet.
  PowerShell lets you access other frameworks and libraries, such as WMI and .NET. Instead of

>   investing time creating a cmdlet, you might be able to get the information more quickly using one

Reviewer: ~~This implies to me as a naive reader that the effort is in _cmdlet development_ as opposed to writing a configuration file and (potentially) parser script. I'm okay if we're using _cmdlet development_ in these documents as shorthand for this, just want to be sure.~~

Edit: Stricken in favor of [this comment]() but left for clarity (and in case the decision is made to update terminology).

```suggestion
  investing time amplifying a command-line tool, you might be able to get the information more quickly using one
```

- **Is the output trivial?**

  In some situations, the command-line tool may produce output that is easy to use in your
  automation. If this is the case, investing time in creating a cmdlet with structured output may
  not be of interest.

## Best Practices

- **Reduce your time of investment by focusing on what you need.**

  When examining a command-line tool for automation, remember you are not required to replicate all
  features of the tool in the new cmdlet. In other words, reduce the time investment by choosing
  only the features or parameters you need to accomplish your goal. If a tool has 12 use cases and

>   you only need the functionality of two of them, create a Crescendo configuration only for the two

Author: https://docs.microsoft.com/en-us/style-guide/numbers
